README.md
# mcresponse | github.com/abbeymart/mcresponse-deno

- Standard return responses for system functional tasks
- It exposes the following function: getResMessage
- See the test folder for test cases / scenarios and usage

## Usage

```ts
import {
  getResMessage,
  Status,
} from "https://deno.land/x/mcresponse@v0.3.1/mod.ts";

// See the test folder for different test cases / scenarios and usage
```


HISTORY.md
# Historical logs

## Version 0.3.2

- Release date: Nov 28, 2023
- Refactor-generic-type-for-getParamsMessage-function-v0.3.2

## Version 0.3.1

- Release date: Nov 27, 2023
- Updated-ValueType-for-consistency-across-packages-v0.3.1

## Version 0.3.0

- Release date: Nov 25, 2023
- Refactored-to-generic-type-v0.3.0

## Version 0.2.1

- Release date: Nov 19, 2022
- Updated-module-version-ref-v0.2.1

## Version 0.2.0

- Release date: Nov 19, 2022
- Activated-for-deno-module-release-v0.2.0

## Version 0.1.0

- Release date: Nov 19, 2022 | refactored to deno module-v0.1.0
- mcresponse package includes the standard responses for various system
  activities/operations
- See the test folder for test cases / scenarios and usage


test.ts
import "./test/checkError.test.ts";
import "./test/success.test.ts";


test/success.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-11
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc: success scenarios testing
 */

import {
  assertEquals,
  assertNotEquals,
  mcTest,
  postTestResult,
} from "../test_deps.ts";
import { getResMessage, Status } from "../src/index.ts";

const msgType = "success",
  res = {
    code: "success",
    resCode: Status.OK,
    resMessage: "OK",
    value: {},
    message: "Request completed successfully",
  };
let options = {
  value: ["a", "b", "c"],
  code: "",
  message: "",
};

(async () => {
  console.log("SUCCESS-TEST-START\n");
  await mcTest({
    name: "should return success code for success-message",
    testFunc: () => {
      const req = getResMessage(msgType, options);
      assertEquals(res.code, req.code, `response-code should be: ${res.code}`);
      assertNotEquals(req.code, "unAuthorized");
    },
  });

  await mcTest({
    name: "should return ok/200 resCode for success-message",
    testFunc: () => {
      const req = getResMessage(msgType);
      assertEquals(
        res.resCode,
        req.resCode,
        `resCode should be: ${res.resCode}`,
      );
      assertEquals(
        res.resMessage,
        req.resMessage,
        `resCode should be: ${res.resMessage}`,
      );
    },
  });

  await mcTest({
    name: "should return Completed successfully message for success-message",
    testFunc: () => {
      const req = getResMessage(msgType, options);
      assertEquals(
        res.message,
        req.message,
        `message should be: ${res.message}`,
      );
    },
  });

  await mcTest({
    name: "should return correct default message",
    testFunc: () => {
      options = {
        value: ["a", "b", "c"],
        code: "",
        message: "completed successfully",
      };
      const req = getResMessage(msgType, options);
      assertEquals(
        req.message.includes(options.message),
        true,
        `response should be: true`,
      );
    },
  });

  await mcTest({
    name: "should return correct custom message",
    testFunc: () => {
      options = {
        value: ["a", "b", "c"],
        code: "",
        message: "custom completed successfully",
      };
      const req = getResMessage("authCode", options);
      assertEquals(req.code, "authCode", `response should be: authCode`);
      assertEquals(
        req.message === options.message,
        true,
        `response should be: true`,
      );
    },
  });

  postTestResult();
  console.log("\nSUCCESS-TEST-END\n");
})();


test/checkError.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-11
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc: check-error testing
 */

import {
  assertEquals,
  assertNotEquals,
  mcTest,
  postTestResult,
} from "../test_deps.ts";
import { getResMessage, Status } from "../src/index.ts";

const msgType = "checkError",
  options = {
    value: {},
    code: "",
    message: "",
  },
  res = {
    code: "paramsError",
    resCode: Status.NotAcceptable,
    resMessage: "Not Acceptable",
    value: "",
    message: "Parameters checking error",
  };

(async () => {
  console.log("CHECK-ERROR-TEST-START\n");
  await mcTest({
    name: "should return paramsError code for checkError-message",
    testFunc: () => {
      const req = getResMessage(msgType, options);
      assertEquals(res.code, req.code, `response-code should be: ${res.code}`);
      assertNotEquals(req.code, "unAuthorized");
    },
  });

  await mcTest({
    name: "should return NOT_ACCEPTABLE/406 resCode",
    testFunc: () => {
      const req = getResMessage(msgType, {});
      assertEquals(
        res.resCode,
        req.resCode,
        `resCode should be: ${res.resCode}`,
      );
      assertEquals(
        res.resMessage,
        req.resMessage,
        `resCode should be: ${res.resMessage}`,
      );
    },
  });

  await mcTest({
    name: "should return Parameters checking error message",
    testFunc: () => {
      const req = getResMessage(msgType, options);
      assertEquals(
        res.message,
        req.message,
        `message should be: ${res.message}`,
      );
    },
  });

  postTestResult();
  console.log("\nCHECK-ERROR-TEST-END\n");
})();


src/resMessage.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-11, 2023-11-25
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: @mconnect/res-messages, response-messages | utility functions
 */

import {
  ValueType,
  MessageObject,
  ResponseMessage,
  ResponseMessageOptions,
  stdResMessages,
} from "./stdResMessages.ts";

function msgFunc<T extends ValueType>(res: {
  code: string;
  resCode: number;
  resMessage: string;
  message: string;
  value: T | ValueType;
}): ResponseMessage<T> {
  return {
    code: res.code,
    message: res.message,
    value: res.value as T,
    resCode: res.resCode,
    resMessage: res.resMessage,
  };
}

export function getResMessage<T extends ValueType>(
  msgType: string,
  options?: ResponseMessageOptions<T>,
): ResponseMessage<T> {
  let value: T | ValueType,
    code: string,
    resCode: number,
    resMessage: string,
    message: string;

  const messageType = msgType || options?.msgType || "unknown";

  const val = stdResMessages[messageType];
  if (val) {
    code = val.code;
    value = options && options.value ? options.value : val.value;
    resCode = val.resCode;
    resMessage = val.resMessage;
    message = options && options.message ? `${options.message}` : val.message;
  } else {
    // use stdResMessages default response
    const val = stdResMessages["unknown"];
    value = options && options.value ? options.value : val.value;
    code = messageType || val.code;
    resCode = val.resCode;
    resMessage = val.resMessage;
    message = options && options.message ? `${options.message}` : val.message;
  }
  return msgFunc<T>({ code, message, value, resCode, resMessage });
}

export function getParamsMessage<T extends ValueType>(
  msgObject: MessageObject,
  msgType = "paramsError",
): ResponseMessage<T> {
  let messages = "";
  for (const [key, msg] of Object.entries(msgObject)) {
    messages = messages ? `${messages} | ${key} : ${msg}` : `${key} : ${msg}`;
  }
  return getResMessage(msgType, {
    message: messages,
  });
}


src/netStatusCode.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-06-25 | @Updated: 2020-07-14
 * @Company: mConnect.biz | @License: MIT
 * @Description: HTTP Status Codes
 */

/** HTTP status codes */
export enum Status {
  /** RFC 7231, 6.2.1 */
  Continue = 100,
  /** RFC 7231, 6.2.2 */
  SwitchingProtocols = 101,
  /** RFC 2518, 10.1 */
  Processing = 102,
  /** RFC 8297 **/
  EarlyHints = 103,
  /** RFC 7231, 6.3.1 */
  OK = 200,
  /** RFC 7231, 6.3.2 */
  Created = 201,
  /** RFC 7231, 6.3.3 */
  Accepted = 202,
  /** RFC 7231, 6.3.4 */
  NonAuthoritativeInfo = 203,
  /** RFC 7231, 6.3.5 */
  NoContent = 204,
  /** RFC 7231, 6.3.6 */
  ResetContent = 205,
  /** RFC 7233, 4.1 */
  PartialContent = 206,
  /** RFC 4918, 11.1 */
  MultiStatus = 207,
  /** RFC 5842, 7.1 */
  AlreadyReported = 208,
  /** RFC 3229, 10.4.1 */
  IMUsed = 226,

  /** RFC 7231, 6.4.1 */
  MultipleChoices = 300,
  /** RFC 7231, 6.4.2 */
  MovedPermanently = 301,
  /** RFC 7231, 6.4.3 */
  Found = 302,
  /** RFC 7231, 6.4.4 */
  SeeOther = 303,
  /** RFC 7232, 4.1 */
  NotModified = 304,
  /** RFC 7231, 6.4.5 */
  UseProxy = 305,
  /** RFC 7231, 6.4.7 */
  TemporaryRedirect = 307,
  /** RFC 7538, 3 */
  PermanentRedirect = 308,

  /** RFC 7231, 6.5.1 */
  BadRequest = 400,
  /** RFC 7235, 3.1 */
  Unauthorized = 401,
  /** RFC 7231, 6.5.2 */
  PaymentRequired = 402,
  /** RFC 7231, 6.5.3 */
  Forbidden = 403,
  /** RFC 7231, 6.5.4 */
  NotFound = 404,
  /** RFC 7231, 6.5.5 */
  MethodNotAllowed = 405,
  /** RFC 7231, 6.5.6 */
  NotAcceptable = 406,
  /** RFC 7235, 3.2 */
  ProxyAuthRequired = 407,
  /** RFC 7231, 6.5.7 */
  RequestTimeout = 408,
  /** RFC 7231, 6.5.8 */
  Conflict = 409,
  /** RFC 7231, 6.5.9 */
  Gone = 410,
  /** RFC 7231, 6.5.10 */
  LengthRequired = 411,
  /** RFC 7232, 4.2 */
  PreconditionFailed = 412,
  /** RFC 7231, 6.5.11 */
  RequestEntityTooLarge = 413,
  /** RFC 7231, 6.5.12 */
  RequestURITooLong = 414,
  /** RFC 7231, 6.5.13 */
  UnsupportedMediaType = 415,
  /** RFC 7233, 4.4 */
  RequestedRangeNotSatisfiable = 416,
  /** RFC 7231, 6.5.14 */
  ExpectationFailed = 417,
  /** RFC 7168, 2.3.3 */
  Teapot = 418,
  /** RFC 7540, 9.1.2 */
  MisdirectedRequest = 421,
  /** RFC 4918, 11.2 */
  UnprocessableEntity = 422,
  /** RFC 4918, 11.3 */
  Locked = 423,
  /** RFC 4918, 11.4 */
  FailedDependency = 424,
  /** RFC 8470, 5.2 */
  TooEarly = 425,
  /** RFC 7231, 6.5.15 */
  UpgradeRequired = 426,
  /** RFC 6585, 3 */
  PreconditionRequired = 428,
  /** RFC 6585, 4 */
  TooManyRequests = 429,
  /** RFC 6585, 5 */
  RequestHeaderFieldsTooLarge = 431,
  /** RFC 7725, 3 */
  UnavailableForLegalReasons = 451,

  /** RFC 7231, 6.6.1 */
  InternalServerError = 500,
  /** RFC 7231, 6.6.2 */
  NotImplemented = 501,
  /** RFC 7231, 6.6.3 */
  BadGateway = 502,
  /** RFC 7231, 6.6.4 */
  ServiceUnavailable = 503,
  /** RFC 7231, 6.6.5 */
  GatewayTimeout = 504,
  /** RFC 7231, 6.6.6 */
  HTTPVersionNotSupported = 505,
  /** RFC 2295, 8.1 */
  VariantAlsoNegotiates = 506,
  /** RFC 4918, 11.5 */
  InsufficientStorage = 507,
  /** RFC 5842, 7.2 */
  LoopDetected = 508,
  /** RFC 2774, 7 */
  NotExtended = 510,
  /** RFC 6585, 6 */
  NetworkAuthenticationRequired = 511,
}

export const StatusText = new Map<Status, string>([
  [Status.Continue, "Continue"],
  [Status.SwitchingProtocols, "Switching Protocols"],
  [Status.Processing, "Processing"],
  [Status.EarlyHints, "Early Hints"],
  [Status.OK, "OK"],
  [Status.Created, "Created"],
  [Status.Accepted, "Accepted"],
  [Status.NonAuthoritativeInfo, "Non-Authoritative Information"],
  [Status.NoContent, "No Content"],
  [Status.ResetContent, "Reset Content"],
  [Status.PartialContent, "Partial Content"],
  [Status.MultiStatus, "Multi-Status"],
  [Status.AlreadyReported, "Already Reported"],
  [Status.IMUsed, "IM Used"],
  [Status.MultipleChoices, "Multiple Choices"],
  [Status.MovedPermanently, "Moved Permanently"],
  [Status.Found, "Found"],
  [Status.SeeOther, "See Other"],
  [Status.NotModified, "Not Modified"],
  [Status.UseProxy, "Use Proxy"],
  [Status.TemporaryRedirect, "Temporary Redirect"],
  [Status.PermanentRedirect, "Permanent Redirect"],
  [Status.BadRequest, "Bad Request"],
  [Status.Unauthorized, "Unauthorized"],
  [Status.PaymentRequired, "Payment Required"],
  [Status.Forbidden, "Forbidden"],
  [Status.NotFound, "Not Found"],
  [Status.MethodNotAllowed, "Method Not Allowed"],
  [Status.NotAcceptable, "Not Acceptable"],
  [Status.ProxyAuthRequired, "Proxy Authentication Required"],
  [Status.RequestTimeout, "Request Timeout"],
  [Status.Conflict, "Conflict"],
  [Status.Gone, "Gone"],
  [Status.LengthRequired, "Length Required"],
  [Status.PreconditionFailed, "Precondition Failed"],
  [Status.RequestEntityTooLarge, "Request Entity Too Large"],
  [Status.RequestURITooLong, "Request URI Too Long"],
  [Status.UnsupportedMediaType, "Unsupported Media Type"],
  [Status.RequestedRangeNotSatisfiable, "Requested Range Not Satisfiable"],
  [Status.ExpectationFailed, "Expectation Failed"],
  [Status.Teapot, "I'm a teapot"],
  [Status.MisdirectedRequest, "Misdirected Request"],
  [Status.UnprocessableEntity, "Unprocessable Entity"],
  [Status.Locked, "Locked"],
  [Status.FailedDependency, "Failed Dependency"],
  [Status.TooEarly, "Too Early"],
  [Status.UpgradeRequired, "Upgrade Required"],
  [Status.PreconditionRequired, "Precondition Required"],
  [Status.TooManyRequests, "Too Many Requests"],
  [Status.RequestHeaderFieldsTooLarge, "Request Header Fields Too Large"],
  [Status.UnavailableForLegalReasons, "Unavailable For Legal Reasons"],
  [Status.InternalServerError, "Internal Server Error"],
  [Status.NotImplemented, "Not Implemented"],
  [Status.BadGateway, "Bad Gateway"],
  [Status.ServiceUnavailable, "Service Unavailable"],
  [Status.GatewayTimeout, "Gateway Timeout"],
  [Status.HTTPVersionNotSupported, "HTTP Version Not Supported"],
  [Status.VariantAlsoNegotiates, "Variant Also Negotiates"],
  [Status.InsufficientStorage, "Insufficient Storage"],
  [Status.LoopDetected, "Loop Detected"],
  [Status.NotExtended, "Not Extended"],
  [Status.NetworkAuthenticationRequired, "Network Authentication Required"],
]);


src/stdResMessages.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-11, 2023-11-25
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: @mconnect/res-messages, response-messages | settings, default values
 */

import { Status, StatusText } from "./netStatusCode.ts";

export type ValueType =
    | Record<string, unknown>
    | Array<Record<string, unknown>>
    | string
    | number
    | Array<string>
    | Array<number>
    | Date
    | Array<Date>
    | boolean
    | Array<boolean>
    | { [key: string]: ValueType }
    | unknown;


export interface ResponseMessage<T extends ValueType> {
    code: string;
    message: string;
    value: T;
    resCode: number;
    resMessage: string;
}

interface MessageParam<T extends ValueType> {
    [key: string]: ResponseMessage<T>;
}

export interface MessageObject {
    [key: string]: string;
}

export interface ResponseMessageOptions<T extends ValueType> {
    msgType?: string;
    message?: string;
    value?: T;
}

export enum MessageCodes {
    paramsError = "paramsError",
    checkError = "checkError",
    connectError = "connectError",
    validateError = "validateError",
    tokenExpired = "tokenExpired",
    unAuthorizedText = "unAuthorized",
    notFoundText = "notFound",
    success = "success",
    removeDenied = "removeDenied",
    removeError = "removeError",
    removed = "removed",
    deleted = "deleted",
    subItems = "subItems",
    duplicate = "duplicate",
    updated = "updated",
    updateError = "updateError",
    updateDenied = "updateDenied",
    inserted = "inserted",
    insertError = "insertError",
    exists = "exists",
    unknown = "unknown",
}

// message options => code, resCode, reMessage, message, value
export const stdResMessages: MessageParam<ValueType> = {
    paramsError  : {
        code      : MessageCodes.paramsError,
        resCode   : Status.NotAcceptable,
        resMessage: StatusText.get(Status.NotAcceptable) || "Not Acceptable",
        message   : "Parameters checking error",
        value     : {},
    },
    checkError   : {
        code      : MessageCodes.paramsError,
        resCode   : Status.NotAcceptable,
        resMessage: StatusText.get(Status.NotAcceptable) || "Not Acceptable",
        message   : "Parameters checking error",
        value     : {},
    },
    connectError : {
        code      : MessageCodes.connectError,
        resCode   : Status.NetworkAuthenticationRequired,
        resMessage: StatusText.get(Status.NetworkAuthenticationRequired) || "Network Auth Required",
        message   : "Connection error",
        value     : {},
    },
    validateError: {
        code      : MessageCodes.paramsError,
        resCode   : Status.NotAcceptable,
        resMessage: StatusText.get(Status.NotAcceptable) || "Not Acceptable",
        message   : "Validation error for inputs parameters",
        value     : {},
    },
    tokenExpired : {
        code      : MessageCodes.tokenExpired,
        resCode   : Status.Unauthorized,
        resMessage: StatusText.get(Status.Unauthorized) || "Not Authorized",
        message   : "Unauthorized. Token / Access-key has expired. Please login again",
        value     : {},
    },
    unAuthorized : {
        code      : MessageCodes.unAuthorizedText,
        resCode   : Status.Unauthorized,
        resMessage: StatusText.get(Status.Unauthorized) || "Not Authorized",
        message   : "Unauthorised Action or Task",
        value     : {},
    },
    notFound     : {
        code      : MessageCodes.notFoundText,
        resCode   : Status.NotFound,
        resMessage: StatusText.get(Status.NotFound) || "Not Found",
        message   : "Requested information not found",
        value     : {},
    },
    success      : {
        code      : MessageCodes.success,
        resCode   : Status.OK,
        resMessage: StatusText.get(Status.OK) || "OK",
        message   : "Request completed successfully",
        value     : {},
    },
    removeDenied : {
        code      : MessageCodes.removeDenied,
        resCode   : Status.Unauthorized,
        resMessage: StatusText.get(Status.Unauthorized) || "Not Authorized",
        message   : "Remove action/task denied/unauthorised",
        value     : {},
    },
    removeError  : {
        code      : MessageCodes.removeError,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Error removing/deleting information, retry or contact system-admin",
        value     : {},
    },
    removed      : {
        code      : MessageCodes.removed,
        resCode   : Status.OK,
        resMessage: StatusText.get(Status.OK) || "OK",
        message   : "Record(s) deleted/removed successfully",
        value     : {},
    },
    subItems     : {
        code      : MessageCodes.subItems,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Record includes sub-items, which must be removed first",
        value     : {},
    },
    duplicateRec : {
        code      : MessageCodes.duplicate,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Duplicate record exists",
        value     : {},
    },
    updated      : {
        code      : MessageCodes.updated,
        resCode   : Status.OK,
        resMessage: StatusText.get(Status.OK) || "OK",
        message   : "Information update action completed successfully",
        value     : {},
    },
    updateError  : {
        code      : MessageCodes.updateError,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Error updating information/record(s)",
        value     : {},
    },
    updateDenied : {
        code      : MessageCodes.updateDenied,
        resCode   : Status.Unauthorized,
        resMessage: StatusText.get(Status.Unauthorized) || "Not Authorized",
        message   : "Update action/task not authorised",
        value     : {},
    },
    inserted     : {
        code      : MessageCodes.inserted,
        resCode   : Status.OK,
        resMessage: StatusText.get(Status.OK) || "OK",
        message   : "Information/record(s) inserted/created successfully",
        value     : {},
    },
    insertError  : {
        code      : MessageCodes.insertError,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Error inserting/creating new information/record",
        value     : {},
    },
    recExist     : {
        code      : MessageCodes.exists,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Document/record exists",
        value     : {},
    },
    unknown      : {
        code      : MessageCodes.unknown,
        resCode   : Status.UnprocessableEntity,
        resMessage: StatusText.get(Status.UnprocessableEntity) || "Not Processed",
        message   : "Unspecified response/error message",
        value     : {},
    },
};
